lib/charms/operator_libs_linux/v0/passwd.py

```diff
@@ -64,604 +45,208 @@
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 2
-
-
-class Error(Exception):
-    """Base class of most errors raised by this library."""
+LIBPATCH = 3
 
-    def __repr__(self):
-        """String representation of the Error class."""
-        return f"<{type(self).__module__}.{type(self).__name__} {self.args}>"
 
-    @property
-    def name(self):
-        """Return a string representation of the model plus class."""
-        return f"<{type(self).__module__}.{type(self).__name__}>"
+def user_exists(user: Union[str, int]) -> Optional[pwd.struct_passwd]:
+    """Check if a user exists.
 
-    @property
-    def message(self):
-        """Return the message passed as an argument."""
-        return self.args[0]
+    Args:
+        user: username or gid of user whose existence to check
 
+    Raises:
+        TypeError: where neither a string or int is passed as the first argument
+    """
+    try:
+        if type(user) is int:
+            return pwd.getpwuid(user)
+        elif type(user) is str:
+            return pwd.getpwnam(user)
+        else:
+            raise TypeError("specified argument '%r' should be a string or int", user)
+    except KeyError:
+        logger.info("specified user '%s' doesn't exist", str(user))
+        return None
 
-class UserState(Enum):
-    """The state of a user on the system or in the cache."""
-
-    Present = "present"
-    Absent = "absent"
-    Disabled = "disabled"
-    NoLogin = "nologin"
 
+def group_exists(group: Union[str, int]) -> Optional[grp.struct_group]:
+    """Check if a group exists.
 
-class UserError(Error):
-    """Raised when there's an error managing a user account."""
+    Args:
+        group: username or gid of user whose existence to check
 
+    Raises:
+        TypeError: where neither a string or int is passed as the first argument
+    """
+    try:
+        if type(group) is int:
+            return grp.getgrgid(group)
+        elif type(group) is str:
+            return grp.getgrnam(group)
+        else:
+            raise TypeError("specified argument '%r' should be a string or int", group)
+    except KeyError:
+        logger.info("specified group '%s' doesn't exist", str(group))
+        return None
+
+
+def add_user(
+    username: str,
+    password: Optional[str] = None,
+    shell: str = "/bin/bash",
+    system_user: bool = False,
+    primary_group: str = None,
+    secondary_groups: List[str] = None,
+    uid: int = None,
+    home_dir: str = None,
+) -> str:
+    """Add a user to the system.
+
+    Will log but otherwise succeed if the user already exists.
```

Comment:
I like this. We could also name this "ensure," if we wanted to be consistent across our interfaces. But I think that "add_user" has an advantage as a name due to being shorter :-)